### ISSUE TITLE
`seastar::engine().current_task()` returns `nullptr`

### ISSUE THREAD

**Issue by Lazin** [scylladb/seastar#1746]
Hi Everyone,
I'm working on something that needs to access `task` objects. I always assumed that the call to `seastar::engine().current_task()` will always return pointer to the current task which runs current pice of code. At least if called properly, from the Seastar shard. But I noticed that sometimes it returns `nullptr`. This happened in unit-tests. The code is doing something like this:

```cpp
co_await seastar::async([] {
    ...
    auto task = seastar::engine().current_task();
    assert(task != nullptr);  // fails 
   ...
});
```

I expected this to work with Seastar's threads as well. Is this expectation incorrect or maybe there are more cases in which `current_task` will return `nullptr`?

**Comment by michoecho** [comment]
Yes, `current_task` returns `nullptr` in seastar threads. See: https://github.com/scylladb/seastar/blob/9a8ae5ddb9a77270b6c3ec1a322f63006d8678dc/src/core/thread.cc#L252

It's because the thread's wakeup task lives on top of the thread's stack, and is destroyed as soon as the thread resumes. The current task pointer becomes invalid, so it is reset to `nullptr` to signal that. AFAIK this is the only place (at least in Seastar itself) where the task is destroyed before it's finished.

I think it could be implemented differently — the thread's wakeup task could live in `thread_context`, not in the thread's stack, and then it would not be destroyed on resume.

But I guess nobody cared about that. In the first place, `current_task`'s main purpose is to generate tasktraces. The case of threads is just special-cased there:
https://github.com/scylladb/seastar/blob/9a8ae5ddb9a77270b6c3ec1a322f63006d8678dc/src/util/backtrace.cc#L157-L161

So I think that if we wanted `current_task` to always be valid, we could make it so.  
Why do you need access to `task` objects, though?

**Comment by Lazin** [comment]
I have two use cases:
1. `task_local_ptr` - non-owning smart pointer that gets invalidated after first scheduling point. It caches pointer to current task in the c-tor and then compares it to current task on every access.
2. Fiber id implementation. Used in my deadlock detector.

I think there is no need to updated Seastar. I can just use `seastar::thread_impl::get()->waiting_task();` if `current_task` returned `nullptr`.
